From 992fefef84c60be1c299d4800d483a0989c1129b Mon Sep 17 00:00:00 2001
From: Afzal Mohammed <afzal@ti.com>
Date: Wed, 25 Jan 2012 20:58:06 +0530
Subject: [PATCH 544/609] PWM: eCAP: Reconfigure only if frequency has changed

Original commit: 83d907e1b05dabc44f3bb64532d7b58d059a14c0

  From git://arago-project.org/git/projects/linux-am33x.git
  And ti-sdk-am335x-evm-05.05.00.00-Linux-x86-Install image

PWM needs to be reconfigured during cpufreq transition
only if input clock frequency has changed. Change notifier
accrodingly.

This prevents flickering during cpufreq transitions. It is
suspected that as PWM frequency is set to a low value,
reconfiguring PWM is causing flicker observable to naked
eye. Irrespective of this fact, this change removes
unnecessary reconfiguration of PWM.

Signed-off-by: Afzal Mohammed <afzal@ti.com>
---
 drivers/pwm/ecap.c |    8 ++++++--
 1 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/pwm/ecap.c b/drivers/pwm/ecap.c
index 3dd6f6d..cc2ed19 100644
--- a/drivers/pwm/ecap.c
+++ b/drivers/pwm/ecap.c
@@ -196,9 +196,13 @@ static int ecap_pwm_request(struct pwm_device *p)
 static int ecap_frequency_transition_cb(struct pwm_device *p)
 {
 	struct ecap_pwm *ep = to_ecap_pwm(p);
-	unsigned long duty_ns;
+	unsigned long duty_ns, rate;
+
+	rate = clk_get_rate(ep->clk);
+	if (rate == p->tick_hz)
+		return 0;
+	p->tick_hz = rate;
 
-	p->tick_hz = clk_get_rate(ep->clk);
 	duty_ns = p->duty_ns;
 	if (pwm_is_running(p)) {
 		pwm_stop(p);
-- 
1.7.5.4

